backoffconnector.go

```diff
@@ -0,0 +1,93 @@
+package discovery
+
+import (
+	"context"
+	lru "github.com/hashicorp/golang-lru"
+	"sync"
+	"time"
+
+	"github.com/libp2p/go-libp2p-core/host"
+	"github.com/libp2p/go-libp2p-core/peer"
+)
+
+// BackoffConnector is a utility to connect to peers, but only if we have not recently tried connecting to them already
+type BackoffConnector struct {
+	cache      *lru.TwoQueueCache
+	host       host.Host
+	connTryDur time.Duration
+	backoff    BackoffFactory
+	mux        sync.Mutex
+}
+
+// NewBackoffConnector creates a utility to connect to peers, but only if we have not recently tried connecting to them already
+// cacheSize is the size of a TwoQueueCache
+// connectionTryDuration is how long we attempt to connect to a peer before giving up
+// backoff describes the strategy used to decide how long to backoff after previously attempting to connect to a peer
+func NewBackoffConnector(h host.Host, cacheSize int, connectionTryDuration time.Duration, backoff BackoffFactory) (*BackoffConnector, error) {
+	cache, err := lru.New2Q(cacheSize)
+	if err != nil {
+		return nil, err
+	}
+
+	return &BackoffConnector{
+		cache:      cache,
+		host:       h,
+		connTryDur: connectionTryDuration,
+		backoff:    backoff,
+	}, nil
+}
+
+type connCacheData struct {
+	nextTry time.Time
+	strat   BackoffStrategy
+}
+
+// Connect attemps to connect to the peers passed in by peerCh. Will not connect to peers if they are within the backoff period.
+func (c *BackoffConnector) Connect(ctx context.Context, peerCh <-chan peer.AddrInfo) {
+	for {
+		select {
+		case pi, ok := <-peerCh:
+			if !ok {
+				return
+			}
+
+			if pi.ID == c.host.ID() || pi.ID == "" {
+				continue
+			}
+
+			c.mux.Lock()
+			val, ok := c.cache.Get(pi.ID)
+			var cachedPeer *connCacheData
+			if ok {
+				tv := val.(*connCacheData)
+				now := time.Now()
+				if now.Before(tv.nextTry) {
+					c.mux.Unlock()
+					continue
+				}
+
+				tv.nextTry = now.Add(tv.strat.Delay())
+			} else {
+				cachedPeer = &connCacheData{strat: c.backoff()}
+				cachedPeer.nextTry = time.Now().Add(cachedPeer.strat.Delay())
+				c.cache.Add(pi.ID, cachedPeer)
+			}
+			c.mux.Unlock()
+
+			go func(pi peer.AddrInfo) {
```

Comment:
I'm concerned that we don't rate limit this at all.

Reply:
Do we need to? In theory the channel could control how many peers it sends in. Also, the dialer should have its own form of rate limiting, no?

Reply:
Kind of.

* The channel will likely return peers as fast as it can find them (if it comes from a discovery service, anyways).
* We will rate-limit inside the dialer (to an extent). But we'll still create a bunch of goroutines.

On the other hand, discovery channels _usually_ limit themselves to N peers. If we can assume that N is <=100, that's probably fine.

Reply:
Added a comment that the caller should be wary of the number and rate of peers being passed into the channel.

Additionally, after talking with @Stebalien it's seeming more and more reasonable that this backoff connector should really live deeper in the libp2p stack so it's existence/usefulness here may be short lived. We'll see though 😄